netket/_exact_dynamics.py

```diff
@@ -0,0 +1,312 @@
+                   other choices are `dense` and `direct`.
+               solver_kwargs (dict): Keyword arguments passed to `solve_ivp`.
+
+           Examples:
+               Solving 1D Ising model with imaginary time propagation:
+
+               ```python
+               >>> import netket as nk
+               >>> import numpy as np
+               >>> L = 8
+               >>> graph = nk.graph.Hypercube(L, n_dim=1, pbc=True)
+               >>> hilbert = nk.hilbert.Spin(graph, 1/2)
+               >>> n_states = hilbert.n_states
+               >>> hamiltonian = nk.operator.Ising(hilbert, h=1.0)
+               >>> psi0 = np.random.rand(n_states)
+               >>> driver = nk.exact.ExactTimePropagation(hamiltonian, stepper, t0=0,
+               ...                                        initial_state=psi0,
+               ...                                        propagation_type="imaginary")
+               >>> driver.add_observable(hamiltonian, 'Hamiltonian')
+               >>> for step in driver.iter(dt=0.05, n_iter=20):
+               ...     obs = driver.get_observable_stats()
+
+               ```
+        """
+        self._h = lambda t: hamiltonian  # FIXME: Support time-dependent H
+        self._rhs = _make_rhs(self._h, propagation_type, matrix_type)
+
+        self.state = initial_state
+        self._dt = dt
+        self._t0 = t0
+        self._steps = 0
+        self._prop_type = propagation_type
+        self._matrix_type = matrix_type
+
+        self._solver_kwargs = {}
+        if solver_kwargs:
+            self._solver_kwargs.update(solver_kwargs)
+
+        # DEPRECATED: Remove in NetKet v3
+        self._obs = {}
+
+    def _estimate_stats(self, obs):
+        if isinstance(obs, _nk.operator.abstract_operator.AbstractOperator):
+            op = _make_op(obs, self._matrix_type)
+        else:
+            op = obs
+
+        v0 = op.dot(self.state)
+        mean = _np.vdot(self.state, v0)
+        variance = _np.vdot(self.state, op.dot(v0)) - mean ** 2
+
+        return _nk.stats.Stats(
+            mean=mean,
+            error_of_mean=0.0,
+            variance=variance.real,
+            correlation=_NaN,
+            R=_NaN,
+        )
+
+    def estimate(self, observables):
+        return _tree_map(self._estimate_stats, observables)
+
+    @property
+    def t(self):
+        """
+        Current simulation time.
+        """
+        return self._t0 + self._dt * self._steps
+
+    @property
+    def dt(self):
+        """
+        Simulation time step.
+        """
+        return self._dt
+
+    @property
+    def step_count(self):
+        return self._step
+
+    def advance(self, n_steps):
+        """
+        Advance the time propagation by `n_steps` simulation steps
+        of duration `self.dt`.
+
+           Args:
+               :n_steps (int): No. of steps to advance.
+        """
+        t_end = self.t + n_steps * self._dt
+        res = _int.solve_ivp(
```

Comment:
I think that efficiency-wise this is rather bad because it will instantiate the full solver at each time step? 
Maybe we should use the [explicit integrators ](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.RK45.html#scipy.integrate.RK45) and just yield the member "y" and the time "t" after calling the method "step" of those?

Reply:
`step` does not actually allow specifying a step size, it will just perform a step with whatever internal `dt` the adaptive solver uses at the time. It is possible to hit a specific time by setting `t_bound`, but once it is reached the solver is set to a `finished` state. Resetting it seems to be not really supported, so I am just doing it manually. It works for now, though it is a bit ugly and I can't guarantee it will work for any type of `OdeSolver`.